benches/coprocessor_executors.rs

```diff
@@ -11,68 +11,512 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+#[macro_use]
 extern crate criterion;
 
+extern crate kvproto;
 extern crate test_coprocessor;
 extern crate tikv;
 extern crate tipb;
 
 use criterion::{black_box, Bencher, Criterion};
 
+use kvproto::coprocessor::KeyRange;
 use tipb::executor::TableScan;
 
 use test_coprocessor::*;
 use tikv::coprocessor::codec::Datum;
 use tikv::coprocessor::dag::executor::{Executor, TableScanExecutor};
+use tikv::storage::RocksEngine;
 
-fn bench_table_scan_primary_key(b: &mut Bencher) {
-    let id = ColumnBuilder::new()
-        .col_type(TYPE_LONG)
-        .primary_key(true)
-        .build();
-    let foo = ColumnBuilder::new().col_type(TYPE_LONG).build();
-    let table = TableBuilder::new()
-        .add_col(id.clone())
-        .add_col(foo.clone())
-        .build();
-
-    let mut store = Store::new();
-    for i in 0..10 {
-        store.begin();
-        store
-            .insert_into(&table)
-            .set(&id, Datum::I64(i))
-            .set(&foo, Datum::I64(0xDEADBEEF))
-            .execute();
-        store.commit();
-    }
-
-    let mut meta = TableScan::new();
-    meta.set_table_id(table.id);
-    meta.set_desc(false);
-    meta.mut_columns().push(id.get_column_info());
-
+fn bench_table_scan_next(
+    b: &mut Bencher,
+    meta: &TableScan,
+    ranges: &[KeyRange],
+    store: &Store<RocksEngine>,
+) {
     b.iter_with_setup(
         || {
             let mut executor = TableScanExecutor::new(
                 meta.clone(),
-                vec![table.get_select_range()],
+                ranges.to_vec(),
                 store.to_fixture_store(),
                 false,
             ).unwrap();
             // There is a step of building scanner in the first `next()` which cost time,
             // so we next() before hand.
-            executor.next().unwrap();
+            executor.next().unwrap().unwrap();
             executor
         },
         |mut executor| {
-            black_box(executor.next().unwrap());
+            black_box(black_box(&mut executor).next().unwrap().unwrap());
```

Comment:
why double `black_box` here?

Reply:
One to mark that `executor` is something that should not be optimized at compile time (because in non-benchmark scenarios it is build dynamically), one to mark that the result is used (because in non-benchmark scenarios the result will be later used to evaluate something else).